File: sorting_&_searching/counting_sort.py

def counting_sort(arr1):
    max_val = max(arr1) + 1
    size = len(arr1)
    count = [0]*max_val
    # output = [0]*size
    output = list()

    # simple count all elements
    for i in range(0, size):
        count[arr1[i]] += 1

    # loop for count cumulative sum
    # for m in range(1, max_val):
    #     count[m] += count[m-1]
    #
    # k = size - 1
    # while k >= 0:
    #     output[count[arr1[k]] - 1] = arr1[k]
    #     count[arr1[k]] -= 1
    #     k -= 1

    for i, val in enumerate(count):
        if val > 0:
            output.extend([i]*val)

    return output


arr = [4, 2, 2, 8, 3, 3, 1]
print(counting_sort(arr))

File: sorting/go_/counting_sort.go
package main

import (
	"fmt"
	"sort"
)

// counting sort time complexity O(m+n)

func CountingSort(arr []int) []int {
	length := len(arr)
	Output := make([]int, length)
	sort.Ints(arr)
	MaxVal := arr[length-1]
	Count := make([]int, MaxVal+1)
	for _, val := range arr {
		Count[val]++
	}
	for i := 1; i < length; i++ {
		Count[i] += Count[i-1]
	}
	i := length - 1
	for i >= 0 {
		val := Count[arr[i]]
		fmt.Println(val, arr[i])
		Output[val-1] = arr[i]
		Count[arr[i]]--
		i--
	}
	return Output
}

func main() {
	arr := []int{4, 2, 2, 8, 3, 3, 1}
	fmt.Println(CountingSort(arr))
}
